substrate/frame/asset-rewards/src/lib.rs

```diff
@@ -0,0 +1,762 @@
+				.or_else(|_| ensure_signed(origin))?;
+
+			ensure!(
+				new_expiry_block > frame_system::Pallet::<T>::block_number(),
+				Error::<T>::ExpiryBlockMustBeInTheFuture
+			);
+
+			let pool_info = Pools::<T>::get(pool_id).ok_or(Error::<T>::NonExistentPool)?;
+			ensure!(pool_info.admin == Some(caller), BadOrigin);
+
+			// Always start by updating the pool rewards.
+			let reward_per_token = Self::reward_per_token(&pool_info)?;
+			let mut pool_info = Self::update_pool_rewards(&pool_info, reward_per_token)?;
+
+			pool_info.expiry_block = new_expiry_block;
+			Pools::<T>::insert(pool_id, pool_info);
+
+			Self::deposit_event(Event::PoolExpiryBlockModified { pool_id, new_expiry_block });
+
+			Ok(())
+		}
+
+		/// Convinience method to deposit reward tokens into a pool.
+		///
+		/// This method is not strictly necessary (tokens could be transferred directly to the
+		/// pool pot address), but is provided for convenience so manual derivation of the
+		/// account id is not required.
+		#[pallet::call_index(7)]
+		pub fn deposit_reward_tokens(
+			origin: OriginFor<T>,
+			pool_id: PoolId,
+			amount: T::Balance,
+		) -> DispatchResult {
+			let caller = ensure_signed(origin)?;
+			let pool_info = Pools::<T>::get(pool_id).ok_or(Error::<T>::NonExistentPool)?;
+			let pool_account_id = Self::pool_account_id(&pool_id)?;
+			T::Assets::transfer(
+				pool_info.reward_asset_id,
+				&caller,
+				&pool_account_id,
+				amount,
+				Preservation::Preserve,
+			)?;
+			Ok(())
+		}
+
+		/// Permissioned method to withdraw reward tokens from a pool.
+		///
+		/// Only the pool admin may perform this operation.
+		#[pallet::call_index(8)]
+		pub fn withdraw_reward_tokens(
+			origin: OriginFor<T>,
+			pool_id: PoolId,
+			amount: T::Balance,
+			dest: T::AccountId,
+		) -> DispatchResult {
+			let caller = T::CreatePoolOrigin::ensure_origin(origin.clone())
+				.or_else(|_| ensure_signed(origin))?;
+
+			let pool_info = Pools::<T>::get(pool_id).ok_or(Error::<T>::NonExistentPool)?;
+			ensure!(pool_info.admin == Some(caller), BadOrigin);
+
+			T::Assets::transfer(
+				pool_info.reward_asset_id,
+				&Self::pool_account_id(&pool_id)?,
+				&dest,
+				amount,
+				// Allow completely draining the account.
+				Preservation::Expendable,
+			)?;
+
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Derive a pool account ID from the pallet's ID.
+		pub fn pool_account_id(id: &PoolId) -> Result<T::AccountId, DispatchError> {
+			if Pools::<T>::contains_key(id) {
+				Ok(T::PalletId::get().into_sub_account_truncating(id))
+			} else {
+				Err(Error::<T>::NonExistentPool.into())
+			}
+		}
+
+		/// Computes update pool and staker reward state.
+		///
+		/// Should be called prior to any operation involving a staker.
+		///
+		/// Returns the updated pool and staker info.
+		///
+		/// NOTE: this function has no side-effects. Side-effects such as storage modifications are
+		/// the responsibility of the caller.
+		pub fn update_pool_and_staker_rewards(
```

Comment:
Immutable version looks better, thanks. 
Take this as a minor note, implement if you like the idea, current version also looks good to me.
If I get everything right, we really calculate two values: `pool.reward_per_token_stored` and `staker.rewards`.
I think we could have just two functions returning these balance values, `derive_reward_per_token_stored` and `derive_rewards`. 
The modification of `PoolInfo` and `PoolStakerInfo` type could be done within a single context (right now you update some part in this function, another part in a call function).
To make it more explicit that `staker.rewards` cannot be updated without `staker. reward_per_token_paid`, or that `pool.reward_per_token_stored ` cannot be update without `pool.last_update_block`, the `PoolInfo` and `PoolStakerInfo` types can have factory methods, for example

``` rust
iml PoolInfo {
  fn update_rewards(old: Self, reward_per_token_stored: Balance) -> Self {
    ... update block number and reward_per_token_stored
  }
}
iml PoolStakerInfo {
  fn update_rewards(old: Self, rewards: Balance, reward_per_token_paid: Balance) -> Self {
    ...
  }
}
```
These factory methods really enforce the correctness when the fields are private, which is not our case. But even with public fields and some docs (recommendation to use the types' methods to update the state or create new) I think it make the rules more explicit.